signature/jws/types.go

```diff
@@ -0,0 +1,113 @@
+package jws
+
+import (
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/notaryproject/notation-core-go/signature"
+)
+
+const (
+	headerKeyAlg                  = "alg"
+	headerKeyCty                  = "cty"
+	headerKeyCrit                 = "crit"
+	headerKeyExpiry               = "io.cncf.notary.expiry"
+	headerKeySigningTime          = "io.cncf.notary.signingTime"
+	headerKeySigningScheme        = "io.cncf.notary.signingScheme"
+	headerKeyAuthenticSigningTime = "io.cncf.notary.authenticSigningTime"
+)
+
+var headerKeys = []string{
+	headerKeyAlg,
+	headerKeyCty,
+	headerKeyCrit,
+	headerKeyExpiry,
+	headerKeySigningTime,
+	headerKeySigningScheme,
+	headerKeyAuthenticSigningTime,
+}
+
+// jwsProtectedHeader contains the set of protected headers.
+type jwsProtectedHeader struct {
+	// Defines which algorithm was used to generate the signature.
+	Algorithm string `json:"alg"`
+
+	// Media type of the secured content (the payload).
+	ContentType string `json:"cty"`
+
+	// Lists the headers that implementation MUST understand and process.
+	Critical []string `json:"crit,omitempty"`
+
+	// The "best by use" time for the artifact, as defined by the signer.
+	Expiry *time.Time `json:"io.cncf.notary.expiry,omitempty"`
+
+	// Specifies the Notary v2 Signing Scheme used by the signature.
+	SigningScheme signature.SigningScheme `json:"io.cncf.notary.signingScheme"`
+
+	// The time at which the signature was generated. only valid when signing scheme is `notary.x509`
+	SigningTime *time.Time `json:"io.cncf.notary.signingTime,omitempty"`
+
+	// The time at which the signature was generated. only valid when signing scheme is `notary.x509.signingAuthority`
+	AuthenticSigningTime *time.Time `json:"io.cncf.notary.authenticSigningTime,omitempty"`
+
+	// The user defined attributes.
+	ExtendedAttributes map[string]interface{} `json:"-"`
+}
+
+// jwsUnprotectedHeader contains the set of unprotected headers.
+type jwsUnprotectedHeader struct {
+	// RFC3161 time stamp token Base64-encoded.
+	TimestampSignature []byte `json:"io.cncf.notary.timestampSignature,omitempty"`
+
+	// List of X.509 Base64-DER-encoded certificates
+	// as defined at https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.6.
+	CertChain [][]byte `json:"x5c"`
+
+	// SigningAgent used for signing
+	SigningAgent string `json:"io.cncf.notary.signingAgent,omitempty"`
+}
+
+// jwsEnvelope is the final Signature envelope.
+type jwsEnvelope struct {
+	// JWSPayload Base64URL-encoded. Raw data should be JSON format.
+	Payload string `json:"payload"`
+
+	// jwsProtectedHeader Base64URL-encoded.
+	Protected string `json:"protected"`
+
+	// Signature metadata that is not integrity Protected
+	Header jwsUnprotectedHeader `json:"header"`
```

Comment:
This should be renamed to Unprotected?

Reply:
According to spec, it should be `header`
```
{
    "payload": "<Base64Url(JWSPayload)>",
    "protected": "<Base64Url(ProtectedHeaders)>",
    "header": {
        "io.cncf.notary.timestamp": "<Base64(TimeStampToken)>",
        "x5c": ["<Base64(DER(leafCert))>", "<Base64(DER(intermediateCACert))>", "<Base64(DER(rootCert))>"]
    },
    "signature": "Base64Url( sign( ASCII( <Base64Url(ProtectedHeader)>.<Base64Url(JWSPayload)> )))"  
}
```